File: frontend/src/types/mission.ts
export type Waypoint = {
  id: string;
  sequence: number;
  latitude: number;
  longitude: number;
  altitudeMeters: number;
  speedMetersPerSecond: number;
};

export type MissionRecord = {
  id: number;
  name: string;
  waypoints: Waypoint[];
  createdAt: string;
  updatedAt: string;
};

export type MissionUploadStatus = {
  missionId: number;
  uploaded: boolean;
  waypointCount: number;
  message: string;
};

export type MissionValidationIssue = {
  code: string;
  waypoint: number | null;
  message: string;
};

export type MissionValidationStatistics = {
  waypoints: number;
  distance: number;
};

export type MissionValidationResult = {
  valid: boolean;
  errors: MissionValidationIssue[];
  warnings: MissionValidationIssue[];
  statistics: MissionValidationStatistics;
};

export type PreFlightCheckStatus = "PASS" | "WARNING" | "FAIL";

export type PreFlightCheck = {
  name: string;
  status: PreFlightCheckStatus;
  mandatory: boolean;
  message: string;
};

export type PreFlightResult = {
  ready: boolean;
  score: number;
  checks: PreFlightCheck[];
  warnings: PreFlightCheck[];
};

File: backend/app/models/mission.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class Waypoint(BaseModel):
    model_config = ConfigDict(frozen=True)

    sequence: int = Field(gt=0)
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)
    altitude_meters: float = Field(gt=0, le=10000)
    speed_meters_per_second: float = Field(gt=0, le=100)


class MissionCreate(BaseModel):
    model_config = ConfigDict(frozen=True)

    name: str = Field(min_length=1, max_length=120)
    waypoints: list[Waypoint] = Field(min_length=1)


class MissionRecord(BaseModel):
    model_config = ConfigDict(frozen=True)

    id: int
    name: str
    waypoints: list[Waypoint]
    created_at: datetime
    updated_at: datetime


class MissionUploadStatus(BaseModel):
    model_config = ConfigDict(frozen=True)

    mission_id: int
    uploaded: bool
    waypoint_count: int
    message: str
